cryodrgn/source.py

```diff
@@ -0,0 +1,400 @@
+                assert (
+                    tgt_indices is None
+                ), "Target indices can only be specified when passing in a preallocated array"
+                tgt_indices = np.arange(len(indices))
+            else:
+                if tgt_indices is not None:
+                    assert len(tgt_indices) == len(
+                        indices
+                    ), "indices/tgt_indices length mismatch"
+                else:
+                    tgt_indices = np.arange(len(indices))
+
+            assert isinstance(tgt_indices, np.ndarray)
+
+            for (index, tgt_index) in zip(indices, tgt_indices):
+                f.seek(self.start)
+                offset = index * self.stride
+                # 'offset' in the call below is w.r.t the current position of f
+                _data = np.fromfile(
+                    f, dtype=self.dtype, count=self.size, offset=offset
+                ).reshape(self.ny, self.nx)
+                data[tgt_index, : self.ny, : self.nx] = _data
+
+            return data
+
+
+class TxtFileSource(ImageSource):
+    def __init__(self, filepath: str, n_workers: int = 1, *args, **kwargs):
+
+        _paths = []
+        for line in open(filepath).readlines():
+            line = line.strip()
+            if not os.path.isabs(line):
+                _paths.append(os.path.join(os.path.dirname(filepath), line))
+            else:
+                _paths.append(line)
+        self.sources = [MRCFileSource(path, *args, **kwargs) for path in _paths]
+
+        # We'll only look at the header from the first .mrcs file, and assume that all headers are compatible
+        header = self.sources[0].header
+        self.ny, self.nx = (
+            header.fields["ny"],
+            header.fields["nx"],
+        )
+        assert self.ny == self.nx, "Only square images supported"
+
+        # Calculate the total length of this source by consulting the individual constituent sources
+        _source_lengths = np.array([s.n for s in self.sources])
+        self.nz = sum(_source_lengths)
+
+        # Maintain a list of (<start>, <end>) tuples indicating the positions of each source
+        _s_i = np.add.accumulate(_source_lengths)
+        self.source_intervals = [(0, _s_i[0])] + [
+            (_s_i[i], _s_i[i + 1]) for i in range(0, len(_s_i) - 1)
+        ]
+        n_workers = min(n_workers, len(self.source_intervals))
+
+        super().__init__(L=self.ny, n=self.nz, n_workers=n_workers, *args, **kwargs)
+
+    def _images(self, indices: np.ndarray):
+        def load_single_mrcs(
+            data: np.ndarray,
+            src: MRCFileSource,
+            src_indices: np.ndarray,
+            tgt_indices: np.ndarray,
+        ):
+            src._images(indices=src_indices, data=data, tgt_indices=tgt_indices)
+
+        data = np.empty((len(indices), self.L, self.L), dtype=self.dtype)
+
+        with futures.ThreadPoolExecutor(self.n_workers) as executor:
```

Comment:
This seems kind of the same as the logic for loading from a star file, no? Can you write a single function that handles both?

Reply:
Can we tackle this as a new issue? There are some other optimizations I can make to the `ImageSource` code (for example issue #229). I'm hesitant to make any changes to that code which does work correctly, though maybe with some redundancies.

Reply:
Opened a new issue on this and assigned to myself. I'll mark this as resolved in here.